<!DOCTYPE html>
<html>
<head>
<meta charset="utf-8">
<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">
<title>Getting started</title>
<style type="text/css">
/**
 * Prism.s theme ported from highlight.js's xcode style
 */
pre code {
  padding: 1em;
}
.token.comment {
  color: #007400;
}
.token.punctuation {
  color: #999;
}
.token.tag,
.token.selector {
  color: #aa0d91;
}
.token.boolean,
.token.number,
.token.constant,
.token.symbol {
  color: #1c00cf;
}
.token.property,
.token.attr-name,
.token.string,
.token.char,
.token.builtin {
  color: #c41a16;
}
.token.inserted {
  background-color: #ccffd8;
}
.token.deleted {
  background-color: #ffebe9;
}
.token.operator,
.token.entity,
.token.url,
.language-css .token.string,
.style .token.string {
  color: #9a6e3a;
}
.token.atrule,
.token.attr-value,
.token.keyword {
  color: #836c28;
}
.token.function,
.token.class-name {
  color: #DD4A68;
}
.token.regex,
.token.important,
.token.variable {
  color: #5c2699;
}
.token.important,
.token.bold {
  font-weight: bold;
}
.token.italic {
  font-style: italic;
}
</style>
<style type="text/css">
body {
  font-family: sans-serif;
  max-width: 800px;
  margin: auto;
  padding: 1em;
  line-height: 1.5;
  box-sizing: border-box;
}
body, .footnotes, code { font-size: .9em; }
li li { font-size: .95em; }
*, *:before, *:after {
  box-sizing: inherit;
}
pre, img { max-width: 100%; }
pre, pre:hover {
  white-space: pre-wrap;
  word-break: break-all;
}
pre code {
  display: block;
  overflow-x: auto;
}
code { font-family: 'DejaVu Sans Mono', 'Droid Sans Mono', 'Lucida Console', Consolas, Monaco, monospace; }
:not(pre) > code, code[class] { background-color: #F8F8F8; }
code.language-undefined, pre > code:not([class]) {
  background-color: inherit;
  border: 1px solid #eee;
}
table {
  margin: auto;
  border-top: 1px solid #666;
}
table thead th { border-bottom: 1px solid #ddd; }
th, td { padding: 5px; }
thead, tfoot, tr:nth-child(even) { background: #eee; }
blockquote {
  color: #666;
  margin: 0;
  padding-left: 1em;
  border-left: 0.5em solid #eee;
}
hr, .footnotes::before { border: 1px dashed #ddd; }
.frontmatter { text-align: center; }
#TOC .numbered li { list-style: none; }
#TOC .numbered { padding-left: 0; }
#TOC .numbered ul { padding-left: 1em; }
table, .body h2 { border-bottom: 1px solid #666; }
.body .appendix, .appendix ~ h2 { border-bottom-style: dashed; }
.footnote-ref a::before { content: "["; }
.footnote-ref a::after { content: "]"; }
.footnotes::before {
  content: "";
  display: block;
  max-width: 20em;
}

@media print {
  body {
    font-size: 12pt;
    max-width: 100%;
  }
  tr, img { page-break-inside: avoid; }
}
@media only screen and (min-width: 992px) {
  pre { white-space: pre; }
}
</style>
</head>
<body>
<div class="include-before">
</div>
<div class="frontmatter">
<div class="title"><h1>Getting started</h1></div>
<div class="author"><h2></h2></div>
<div class="date"><h3></h3></div>
</div>
<div class="body">
<h1 id="spacedeconv">spacedeconv</h1>
<p>spacedeconv is a unified interface to 31 deconvolution tools with focus on spatial transcriptomics datasets. In total 17 second-generation deconvolution tools are included, enabling deconvolution of any cell types when single-cell reference data is available. Additionally 10 first-generation tools, which are focussing on deconvolution of immune cells, are available as well as 4 first-generation methods optimised for mouse data. These methods donâ€™t require scRNA-seq data to perform deconvolution as they utilize precomputed and verified signatures.</p>
<h1 id="installation">Installation</h1>
<p>There are two ways to install <code>spacedeconv</code>:</p>
<ul>
<li>The <em>minimal</em> installation installs only the dependencies required for the basic functionalities. All deconvolution methods need to be installed on-demand.</li>
<li>The <em>complete</em> installtation installs all dependencies including all deconvolution methods. This may take a considerable time.</li>
</ul>
<p>Since not all dependencies are on CRAN or Bioconductor, <code>spacedeconv</code> is available from GitHub only. We recommend installing trough the pak package manager:</p>
<pre><code class="language-r">install.packages(&quot;pak&quot;)

# minimal installation
pak::pkg_install(&quot;omnideconv/spacedeconv&quot;)

# complete installation
pak::pkg_install(&quot;omnideconv/spacedeconv&quot;, dependencies = TRUE)
</code></pre>
<h1 id="data-requirements">Data requirements</h1>
<p>spacedeconv offers convenient access to perform first- and second-generation deconvolution on spatial transcriptomics datasets. While deconvolution can be performed directly with first-generation methods, second-generation algorithms require an additional annotated single-cell reference.</p>
<ul>
<li><em><a href="https://bioconductor.org/packages/release/bioc/vignettes/SpatialExperiment/inst/doc/SpatialExperiment.html">SpatialExperiment</a></em>, will be deconvoluted</li>
<li><em><a href="https://bioconductor.org/packages/release/bioc/vignettes/SingleCellExperiment/inst/doc/intro.html">SingleCellExperiment</a></em> (recommended), <em><a href="https://anndata.dynverse.org/">anndata</a></em> or <em><a href="https://satijalab.org/seurat/">Seurat</a></em> containing cell type information</li>
</ul>
<p>You can load 10X Visium Data by providing the <a href="https://support.10xgenomics.com/spatial-gene-expression/software/pipelines/latest/what-is-space-ranger">spaceranger</a> output folder. It is further possible to run spacedeconv with manually created SpatialExperiments. See the SpatialExperiment <a href="https://bioconductor.org/packages/release/bioc/vignettes/SpatialExperiment/inst/doc/SpatialExperiment.html">Documentation</a> for further details.</p>
<pre><code class="language-r">spe &lt;- SpatialExperiment::read10xVisium(&quot;path_to_directory&quot;)
</code></pre>
<h1 id="workflow">Workflow</h1>
<p>The core modules of spacedeconv are used to build a reference signature and deconvolute spatial transcriptomics data. We further provide convenient functions for normalization and visualization. A full list of deconvolution algorithms can be accessed by <code>spacedeconv::deconvolution_methods</code> or in the <a href="spacedeconv_faq.html">FAQ</a>.</p>
<h3 id="1-normalization">1. Normalization</h3>
<p>spacedeconv offers a function to normalize spatial expression data. The normalization is saved in an additional assay, so make sure to use the correct data during deconvolution by providing the desired assay with the parameters <code>assay_sc</code> and <code>assay_sp</code>.</p>
<pre><code class="language-r">spe &lt;- spacedeconv::normalize(spe, method = &quot;cpm&quot;)

# specify expression assay to use
signature &lt;- spacedeconv::build_model(spe,
  method = &quot;quantiseq&quot;,
  assay_sp = &quot;cpm&quot;
)
</code></pre>
<h3 id="2-building-a-reference-signature">2. Building a reference Signature</h3>
<p>All second-generation deconvolution tools require a cell-type specific reference signature which can be calculated with <code>build_model()</code> function. To build a signature annotated single-cell sequencing data is required.</p>
<pre><code class="language-r">signature &lt;- spacedeconv::build_model(
  single_cell_object,
  cell_type_col = &quot;celltype_major&quot;,
  method = &quot;spotlight&quot;,
  assay_sc = &quot;cpm&quot;
)
</code></pre>
<h3 id="3-deconvolution">3. Deconvolution</h3>
<p>This function deconvolutes spatial transcriptomics dataset using one of spacedeconvs 31 deconvolution tools. Since first-generation deconvolution tools contain predefined expression signatures these tools can be used directly on the spatial data. You do not need to provide a signature for this case. spacedeconv returns a SpatialExperiment object annotated with deconvolution results.</p>
<pre><code class="language-r"># save the results to an annotated SpatialExperiment
result &lt;- spacedeconv::deconvolute(
  spatial_object,
  signature,
  method = &quot;spotlight&quot;
)

# return deconvolution results in table form
result &lt;- spacedeconv::deconvolute(
  spatial_object,
  signature,
  method = &quot;spotlight&quot;,
  return_object = FALSE
)
</code></pre>
<h3 id="4-visualization">4. Visualization</h3>
<p>We offer 6 visualization function aiming different insight of the data. A full example of each visualization option is available in the visualization <a href="spacedeconv_visualization.html">Vignette</a>.</p>
<pre><code class="language-r"># sample does refer to the first column of ColData(spe)
# for cell_type input a celltype present in the deconvolution result
spacedeconv::plot_celltype(spe, cell_type = &quot;spotlight_B.cells&quot;)

# umi count
spacedeconv::plot_umi_count(spe)
</code></pre>
</div>
<div class="include-after">
</div>
<script src="https://cdn.jsdelivr.net/npm/prismjs@1.29.0/components/prism-core.min.js" defer></script>
<script src="https://cdn.jsdelivr.net/npm/prismjs@1.29.0/plugins/autoloader/prism-autoloader.min.js" defer></script>
</body>
</html>
